extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -475,39 +475,26 @@ export class Bind {
         this.addMacros_(),
         this.addBindingsForNodes_([root]),
       ]);
-    }).then(results => {
-      dev().info(TAG, '⤷', 'Δ:', results);
+    }).then(() => {
       // Listen for DOM updates (e.g. template render) to rescan for bindings.
       root.addEventListener(AmpEvents.DOM_UPDATE, e => this.onDomUpdate_(e));
       // In dev mode, check default values against initial expression results.
       if (getMode().development) {
         return this.evaluate_().then(results => this.verify_(results));
       }
-    }).then(() => this.checkReadiness_(root));
-  }
-
-  /**
-   * Bind is "ready" when its initialization completes _and_ all <amp-state>
-   * elements' local data is parsed and processed (not remote data).
-   * @param {!Node} root
-   * @private
-   */
-  checkReadiness_(root) {
-    const ampStates = root.querySelectorAll('AMP-STATE');
-    if (ampStates.length > 0) {
-      const whenBuilt = toArray(ampStates).map(el => el.whenBuilt());
-      Promise.all(whenBuilt).then(() => this.onReady_());
-    } else {
-      this.onReady_();
-    }
-  }
-
-  /**
-   * @private
-   */
-  onReady_() {
-    this.viewer_.sendMessage('bindReady', undefined);
-    this.dispatchEventForTesting_(BindEvents.INITIALIZE);
+    }).then(() => {
+      const ampStates = root.querySelectorAll('AMP-STATE');
+      // Force all query-able <amp-state> elements to be built.
+      // A bit weird but amp-state is just metadata, so there's no point
+      // to wait for runtime to schedule builds before sending "bindReady".
+      const whenBuilt = toArray(ampStates).map(el => el.build());
```

Comment:
This doesn't synchronize build state with resource state, build has to come down from resources. I'm a bit confused why we need to force build, though, or why amp-state can't be a prerenderable element itself.

Reply:
Changed this to be an API on amp-state element itself instead as discussed.